### DISCUSSION TITLE
What is the `referrer` argument to `RefResolver`?

### DISCUSSION THREAD

**Discussion by mdavis-xyz** [python-jsonschema/jsonschema#1044] [Q&A] [answered]
What should the `referrer` argument to `RefResolver` be? 

[The documentation](https://python-jsonschema.readthedocs.io/en/stable/api/jsonschema/validators/#jsonschema.validators.RefResolver) just says "The actual referring document". What does that mean? Is that:

* a string of the yaml document
* a dictionary of the yaml document
* a string of the relative file path: `./spec.yaml`
* a string of the full file path: `/path/to/spec.yaml`
* a URI for the file path: `file:///path/to/spec.yaml`
* any of the above, but for json not yaml
* a `pathlib.Path` of the file path

The example in the FAQ uses `referrer=True`. How is `True` an instance of an `actual referring document`?

I looked at the code implementing `RefResolver`. It gets used like `if document is self.referrer:`. How can that work if `referrer` is True? I'm confused.

**Comment by Julian** [comment, marked as answer]
This is going to change in the next version when the [referencing](https://referencing.readthedocs.io/en/stable/) library replaces `RefResolver`.

But it's the actual dictionary it sounds like in your case. In "official" JSON Schema dialects, schema documents are either mappings (dictionaries in Python) or boolean schemas, so in theory a referrer of `True` is "sensical". The referrer is used for fragments / anchors / JSON Pointers -- e.g. for dereferencing `#` or `#/foo/bar` -- a referrer of `True` would contain no anchors (it contains no properties at all of course) so it should never hit code paths that dereference an anchor within but it's indeed possible to have a schema `true` be the root document for a ref resolver and to uselessly write `#` and get back the whole schema or whatever.

But yeah these interfaces are old and not easy to understand, as I say, this should be easier with `referencing` in the next version (as soon as I finish the work actually *deprecating* `RefResolver`, which is the last remaining step).

**Reply by mdavis-xyz** [reply]
Ok, so what does `referrer=False` do?

**Reply by Julian** [reply]
It creates a RefResolver whose schema is the boolean schema `false`.

**Reply by mdavis-xyz** [reply]
Are you talking about [this](http://json-schema.org/draft/2020-12/json-schema-core.html#name-boolean-json-schemas)?

Sorry I still don't understand. I thought the RefResolver is used to look up something like `#/foo/bar` from a subset of a schema, and the thing it looks it up from is `referrer`, as the parent/root schema. `True` is a schema that marks every possible object as valid. But how does that get used when trying to find `True['foo']['bar']`? Are you saying that all anchor lookups will fail, because `True` is really `{}` which contains no `foo` member?

**Reply by Julian** [reply]
> Are you talking about [this](http://json-schema.org/draft/2020-12/json-schema-core.html#name-boolean-json-schemas)?

Yep exactly!

> Are you saying that all anchor lookups will fail, because True is really {} which contains no foo member?

Yes exactly, your understanding is correct it sounds like -- all lookups will fail -- not really because `True` is `{}`, that's a "conceptual" equality -- `True["foo"]` does not exist / is a TypeError in Python / is undefined as a JSON Pointer. That's what I meant by making `referrer` be those values is "sensical" but not "sensible". And the same for `false`. Does that help?

**Comment by mdavis-xyz** [comment]
Yes that helps.

Should I write a PR to clarify the documentation, or is the upcoming change so imminent or substantial that it's not worth bothering?

**Reply by Julian** [reply]
Great! I'm more than happy to review a PR but the referencing change is happening on [this branch](https://github.com/python-jsonschema/jsonschema/tree/referencing) -- it's coming along, I expect it to be done in the next 2 weeks or so (and to be in the next release) which will obsolete `RefResolver` entirely, so it's likely not worth bothering given it probably will never see the light of day anyhow. Feedback will definitely be welcome on the new API though once a beta is released!

**Reply by mdavis-xyz** [reply]
Ah ok. Never mind then. If someone else has the same question as me in the next two weeks, hopefully they can just find this thread.
